TS2322: inferred string instead value type since TS3.2

<!-- 🚨 STOP 🚨 𝗦𝗧𝗢𝗣 🚨 𝑺𝑻𝑶𝑷 🚨

Half of all issues filed here are duplicates, answered in the FAQ, or not appropriate for the bug tracker. Even if you think you've found a *bug*, please read the FAQ first, especially the Common "Bugs" That Aren't Bugs section!

Please help us by doing the following steps before logging an issue:
  * Search: https://github.com/Microsoft/TypeScript/search?type=Issues
  * Read the FAQ: https://github.com/Microsoft/TypeScript/wiki/FAQ

Please fill in the *entire* template below.
-->

<!-- Please try to reproduce the issue with `typescript@next`. It may have already been fixed. -->
**TypeScript Version:** 

Errors: `3.3.0-dev.20181204`, `3.2.1`
Works: `3.1.6`

<!-- Search terms you tried before logging this (so others can find this issue more easily) -->
**Search Terms:**

3.2, SFC, React, default, params, value type, string


**Code**

```ts
import * as React from 'react';

type AorB = 'a' | 'b';

type Props1 = {
  aOrB: AorB;
}

type Props2 = {
  aOrB?: AorB;
}

export const Comp1: React.SFC<Props1> = () => null;

export const Comp2: React.SFC<Props2> = ({ aOrB = 'a' }) => <Comp1 aOrB={aOrB} />;
```

```
{
  "compilerOptions": {
    "jsx": "react",
    "noEmit": true
  }
}
```

```
{
  "name": "ts",
  "version": "1.0.0",
  "license": "MIT",
  "scripts": {
    "type-check": "tsc"
  },
  "devDependencies": {
    "@types/react": "^16.7.13",
    "typescript": "3.x.x"
  }
}
```

**Expected behavior:**

For `3.1`:

```
$ yarn type-check
yarn run v1.12.3
$ tsc
✨  Done in 3.62s.
```

**Actual behavior:**

For `>=3.2`:

```
$ yarn type-check
yarn run v1.12.3
$ tsc
src/index.tsx:15:68 - error TS2322: Type 'string' is not assignable to type 'AorB'.

15 export const Comp2: React.SFC<Props2> = ({ aOrB = 'a' }) => <Comp1 aOrB={aOrB} />;
                                                                      ~~~~

  src/index.tsx:6:3
    6   aOrB: AorB;
        ~~~~
    The expected type comes from property 'aOrB' which is declared here on type 'IntrinsicAttributes & Props1 & { children?: ReactNode; }'


Found 1 error.

error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

**Playground Link:** <!-- A link to a TypeScript Playground "Share" link which demonstrates this behavior -->

**Related Issues:** <!-- Did you find other bugs that looked similar? -->

The above snippet compiled successfully in `3.2.0-dev.20181011` but failed in `3.2.0-dev.20181017`.
I think the related changes are introduced between these two versions.

Note: we can also use this smaller snippet to reproduce the same error without depending on `@types/react`.
This snippet also compiles successfully in `3.2.0-dev.20181011` but fails in `3.2.0-dev.20181017`.

```typescript
type MyFunc<P> = (x: P) => void


interface Props {
  n?: 'one' | 'two' | 'three'
}

function oneTwoThree(n: 'one'|'two'|'three') {
  console.log(n)
}

const hello: MyFunc<Props> = ({n = 'one'}) => {
  oneTwoThree(n);
}

hello({});
```
```json
{
  "compilerOptions": {
    "target": "esnext",
    "module": "commonjs",
    "lib": ["esnext"],
    "noEmit": true,
    "strict": true
  }
}

```
```
index.ts:12:15 - error TS2345: Argument of type 'string' is not assignable to parameter of type '"one" | "two" | "three"'.

12   oneTwoThree(n);
                 ~
```

--- a/src/compiler/checker.ts
+++ b/src/compiler/checker.ts
@@ -4849,7 +4849,7 @@ namespace ts {
             if (strictNullChecks && declaration.initializer && !(getFalsyFlags(checkDeclarationInitializer(declaration)) & TypeFlags.Undefined)) {
                 type = getTypeWithFacts(type, TypeFacts.NEUndefined);
             }
-            return declaration.initializer && !getEffectiveTypeAnnotationNode(walkUpBindingElementsAndPatterns(declaration)) ?
+            return declaration.initializer && !getEffectiveTypeAnnotationNode(walkUpBindingElementsAndPatterns(declaration)) && !getContextualTypeForVariableLikeDeclaration(walkUpBindingElementsAndPatterns(declaration)) ?
                 getUnionType([type, checkDeclarationInitializer(declaration)], UnionReduction.Subtype) :
                 type;
         }
@@ -17020,6 +17020,32 @@ namespace ts {
             }
         }
 
+        function getContextualTypeForVariableLikeDeclaration(declaration: VariableLikeDeclaration): Type | undefined {
+            const typeNode = getEffectiveTypeAnnotationNode(declaration);
+            if (typeNode) {
+                return getTypeFromTypeNode(typeNode);
+            }
+            switch (declaration.kind) {
+                case SyntaxKind.Parameter:
+                    return getContextuallyTypedParameterType(declaration);
+                case SyntaxKind.BindingElement:
+                    return getContextualTypeForBindingElement(declaration);
+                // By default, do nothing and return undefined - only parameters and binding elements have context implied by a parent
+            }
+        }
+
+        function getContextualTypeForBindingElement(declaration: BindingElement): Type | undefined {
+            const parentDeclaration = declaration.parent.parent;
+            const name = declaration.propertyName || declaration.name;
+            const parentType = getContextualTypeForVariableLikeDeclaration(parentDeclaration);
+            if (parentType && !isBindingPattern(name)) {
+                const text = getTextOfPropertyName(name);
+                if (text !== undefined) {
+                    return getTypeOfPropertyOfType(parentType, text);
+                }
+            }
+        }
+
         // In a variable, parameter or property declaration with a type annotation,
         //   the contextual type of an initializer expression is the type of the variable, parameter or property.
         // Otherwise, in a parameter declaration of a contextually typed function expression,
@@ -17031,31 +17057,12 @@ namespace ts {
         function getContextualTypeForInitializerExpression(node: Expression): Type | undefined {
             const declaration = <VariableLikeDeclaration>node.parent;
             if (hasInitializer(declaration) && node === declaration.initializer) {
-                const typeNode = getEffectiveTypeAnnotationNode(declaration);
-                if (typeNode) {
-                    return getTypeFromTypeNode(typeNode);
+                const result = getContextualTypeForVariableLikeDeclaration(declaration);
+                if (result) {
+                    return result;
                 }
-                if (declaration.kind === SyntaxKind.Parameter) {
-                    const type = getContextuallyTypedParameterType(declaration);
-                    if (type) {
-                        return type;
-                    }
-                }
-                if (isBindingPattern(declaration.name)) {
+                if (isBindingPattern(declaration.name)) { // This is less a contextual type and more an implied shape - in some cases, this may be undesirable
                     return getTypeFromBindingPattern(declaration.name, /*includePatternInType*/ true, /*reportErrors*/ false);
-                }
-                if (isBindingPattern(declaration.parent)) {
-                    const parentDeclaration = declaration.parent.parent;
-                    const name = (declaration as BindingElement).propertyName || declaration.name;
-                    if (parentDeclaration.kind !== SyntaxKind.BindingElement) {
-                        const parentTypeNode = getEffectiveTypeAnnotationNode(parentDeclaration);
-                        if (parentTypeNode && !isBindingPattern(name)) {
-                            const text = getTextOfPropertyName(name);
-                            if (text) {
-                                return getTypeOfPropertyOfType(getTypeFromTypeNode(parentTypeNode), text);
-                            }
-                        }
-                    }
                 }
             }
             return undefined;

--- a/src/compiler/checker.ts
+++ b/src/compiler/checker.ts
@@ -4849,7 +4849,7 @@ namespace ts {
             if (strictNullChecks && declaration.initializer && !(getFalsyFlags(checkDeclarationInitializer(declaration)) & TypeFlags.Undefined)) {
                 type = getTypeWithFacts(type, TypeFacts.NEUndefined);
             }
-            return declaration.initializer && !getEffectiveTypeAnnotationNode(walkUpBindingElementsAndPatterns(declaration)) && !getContextualTypeForVariableLikeDeclaration(walkUpBindingElementsAndPatterns(declaration)) ?
+            return declaration.initializer && !getContextualTypeForVariableLikeDeclaration(walkUpBindingElementsAndPatterns(declaration)) ?
                 getUnionType([type, checkDeclarationInitializer(declaration)], UnionReduction.Subtype) :
                 type;
         }

--- a/tests/baselines/reference/destructuringArrayBindingPatternAndAssignment3.types
+++ b/tests/baselines/reference/destructuringArrayBindingPatternAndAssignment3.types
@@ -45,8 +45,8 @@ const [f, g = f, h = i, i = f] = [1]; // error for h = i
 >c : number
 >d : number
 >c : number
->e : any
->e : any
+>e : number
+>e : number
 
 })([1]);
 >[1] : number[]

--- a/tests/baselines/reference/destructuringInitializerContextualTypeFromContext.js
+++ b/tests/baselines/reference/destructuringInitializerContextualTypeFromContext.js
@@ -1,0 +1,50 @@
+//// [destructuringInitializerContextualTypeFromContext.ts]
+interface SFC<P = {}> {
+    (props: P & { children?: any }): any | null;
+}
+
+interface Props {
+    name: "Apollo" | "Artemis" | "Dionysus" | "Persephone";
+}
+
+const Parent: SFC<Props> = ({
+    children,
+    name = "Artemis",
+    ...props
+}) => Child({name, ...props});
+
+const Child: SFC<Props> = ({
+    children,
+    name = "Artemis",
+    ...props
+}) => `name: ${name} props: ${JSON.stringify(props)}`;
+
+//// [destructuringInitializerContextualTypeFromContext.js]
+var __assign = (this && this.__assign) || function () {
+    __assign = Object.assign || function(t) {
+        for (var s, i = 1, n = arguments.length; i < n; i++) {
+            s = arguments[i];
+            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
+                t[p] = s[p];
+        }
+        return t;
+    };
+    return __assign.apply(this, arguments);
+};
+var __rest = (this && this.__rest) || function (s, e) {
+    var t = {};
+    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
+        t[p] = s[p];
+    if (s != null && typeof Object.getOwnPropertySymbols === "function")
+        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) if (e.indexOf(p[i]) < 0)
+            t[p[i]] = s[p[i]];
+    return t;
+};
+var Parent = function (_a) {
+    var children = _a.children, _b = _a.name, name = _b === void 0 ? "Artemis" : _b, props = __rest(_a, ["children", "name"]);
+    return Child(__assign({ name: name }, props));
+};
+var Child = function (_a) {
+    var children = _a.children, _b = _a.name, name = _b === void 0 ? "Artemis" : _b, props = __rest(_a, ["children", "name"]);
+    return "name: " + name + " props: " + JSON.stringify(props);
+};

--- a/tests/baselines/reference/destructuringInitializerContextualTypeFromContext.symbols
+++ b/tests/baselines/reference/destructuringInitializerContextualTypeFromContext.symbols
@@ -1,0 +1,58 @@
+=== tests/cases/compiler/destructuringInitializerContextualTypeFromContext.ts ===
+interface SFC<P = {}> {
+>SFC : Symbol(SFC, Decl(destructuringInitializerContextualTypeFromContext.ts, 0, 0))
+>P : Symbol(P, Decl(destructuringInitializerContextualTypeFromContext.ts, 0, 14))
+
+    (props: P & { children?: any }): any | null;
+>props : Symbol(props, Decl(destructuringInitializerContextualTypeFromContext.ts, 1, 5))
+>P : Symbol(P, Decl(destructuringInitializerContextualTypeFromContext.ts, 0, 14))
+>children : Symbol(children, Decl(destructuringInitializerContextualTypeFromContext.ts, 1, 17))
+}
+
+interface Props {
+>Props : Symbol(Props, Decl(destructuringInitializerContextualTypeFromContext.ts, 2, 1))
+
+    name: "Apollo" | "Artemis" | "Dionysus" | "Persephone";
+>name : Symbol(Props.name, Decl(destructuringInitializerContextualTypeFromContext.ts, 4, 17))
+}
+
+const Parent: SFC<Props> = ({
+>Parent : Symbol(Parent, Decl(destructuringInitializerContextualTypeFromContext.ts, 8, 5))
+>SFC : Symbol(SFC, Decl(destructuringInitializerContextualTypeFromContext.ts, 0, 0))
+>Props : Symbol(Props, Decl(destructuringInitializerContextualTypeFromContext.ts, 2, 1))
+
+    children,
+>children : Symbol(children, Decl(destructuringInitializerContextualTypeFromContext.ts, 8, 29))
+
+    name = "Artemis",
+>name : Symbol(name, Decl(destructuringInitializerContextualTypeFromContext.ts, 9, 13))
+
+    ...props
+>props : Symbol(props, Decl(destructuringInitializerContextualTypeFromContext.ts, 10, 21))
+
+}) => Child({name, ...props});
+>Child : Symbol(Child, Decl(destructuringInitializerContextualTypeFromContext.ts, 14, 5))
+>name : Symbol(name, Decl(destructuringInitializerContextualTypeFromContext.ts, 12, 13))
+>props : Symbol(props, Decl(destructuringInitializerContextualTypeFromContext.ts, 10, 21))
+
+const Child: SFC<Props> = ({
+>Child : Symbol(Child, Decl(destructuringInitializerContextualTypeFromContext.ts, 14, 5))
+>SFC : Symbol(SFC, Decl(destructuringInitializerContextualTypeFromContext.ts, 0, 0))
+>Props : Symbol(Props, Decl(destructuringInitializerContextualTypeFromContext.ts, 2, 1))
+
+    children,
+>children : Symbol(children, Decl(destructuringInitializerContextualTypeFromContext.ts, 14, 28))
+
+    name = "Artemis",
+>name : Symbol(name, Decl(destructuringInitializerContextualTypeFromContext.ts, 15, 13))
+
+    ...props
+>props : Symbol(props, Decl(destructuringInitializerContextualTypeFromContext.ts, 16, 21))
+
+}) => `name: ${name} props: ${JSON.stringify(props)}`;
+>name : Symbol(name, Decl(destructuringInitializerContextualTypeFromContext.ts, 15, 13))
+>JSON.stringify : Symbol(JSON.stringify, Decl(lib.es5.d.ts, --, --), Decl(lib.es5.d.ts, --, --))
+>JSON : Symbol(JSON, Decl(lib.es5.d.ts, --, --), Decl(lib.es5.d.ts, --, --))
+>stringify : Symbol(JSON.stringify, Decl(lib.es5.d.ts, --, --), Decl(lib.es5.d.ts, --, --))
+>props : Symbol(props, Decl(destructuringInitializerContextualTypeFromContext.ts, 16, 21))
+

--- a/tests/baselines/reference/destructuringInitializerContextualTypeFromContext.types
+++ b/tests/baselines/reference/destructuringInitializerContextualTypeFromContext.types
@@ -1,0 +1,57 @@
+=== tests/cases/compiler/destructuringInitializerContextualTypeFromContext.ts ===
+interface SFC<P = {}> {
+    (props: P & { children?: any }): any | null;
+>props : P & { children?: any; }
+>children : any
+>null : null
+}
+
+interface Props {
+    name: "Apollo" | "Artemis" | "Dionysus" | "Persephone";
+>name : "Apollo" | "Artemis" | "Dionysus" | "Persephone"
+}
+
+const Parent: SFC<Props> = ({
+>Parent : SFC<Props>
+>({    children,    name = "Artemis",    ...props}) => Child({name, ...props}) : ({ children, name, ...props }: Props & { children?: any; }) => any
+
+    children,
+>children : any
+
+    name = "Artemis",
+>name : "Apollo" | "Artemis" | "Dionysus" | "Persephone"
+>"Artemis" : "Artemis"
+
+    ...props
+>props : {}
+
+}) => Child({name, ...props});
+>Child({name, ...props}) : any
+>Child : SFC<Props>
+>{name, ...props} : { name: "Apollo" | "Artemis" | "Dionysus" | "Persephone"; }
+>name : "Apollo" | "Artemis" | "Dionysus" | "Persephone"
+>props : {}
+
+const Child: SFC<Props> = ({
+>Child : SFC<Props>
+>({    children,    name = "Artemis",    ...props}) => `name: ${name} props: ${JSON.stringify(props)}` : ({ children, name, ...props }: Props & { children?: any; }) => string
+
+    children,
+>children : any
+
+    name = "Artemis",
+>name : "Apollo" | "Artemis" | "Dionysus" | "Persephone"
+>"Artemis" : "Artemis"
+
+    ...props
+>props : {}
+
+}) => `name: ${name} props: ${JSON.stringify(props)}`;
+>`name: ${name} props: ${JSON.stringify(props)}` : string
+>name : "Apollo" | "Artemis" | "Dionysus" | "Persephone"
+>JSON.stringify(props) : string
+>JSON.stringify : { (value: any, replacer?: (key: string, value: any) => any, space?: string | number): string; (value: any, replacer?: (string | number)[], space?: string | number): string; }
+>JSON : JSON
+>stringify : { (value: any, replacer?: (key: string, value: any) => any, space?: string | number): string; (value: any, replacer?: (string | number)[], space?: string | number): string; }
+>props : {}
+

--- a/tests/baselines/reference/optionalParameterInDestructuringWithInitializer.types
+++ b/tests/baselines/reference/optionalParameterInDestructuringWithInitializer.types
@@ -167,7 +167,7 @@ function func7( {a: {b, c = 6} = {b: 4, c: 5}, d}: {a: {b: number, c?: number}, 
 >b : number
 >c : number
 >6 : 6
->{b: 4, c: 5} : { b: number; c?: number; }
+>{b: 4, c: 5} : { b: number; c: number; }
 >b : number
 >4 : 4
 >c : number

--- a/tests/baselines/reference/sourceMapValidationDestructuringParameterNestedObjectBindingPatternDefaultValues.types
+++ b/tests/baselines/reference/sourceMapValidationDestructuringParameterNestedObjectBindingPatternDefaultValues.types
@@ -50,7 +50,7 @@ function foo1(
 >"secondary" : "secondary"
 
         } = { primary: "SomeSkill", secondary: "someSkill" }
->{ primary: "SomeSkill", secondary: "someSkill" } : { primary?: string; secondary?: string; }
+>{ primary: "SomeSkill", secondary: "someSkill" } : { primary: string; secondary: string; }
 >primary : string
 >"SomeSkill" : "SomeSkill"
 >secondary : string
@@ -88,7 +88,7 @@ function foo2(
 >"secondary" : "secondary"
 
         } = { primary: "SomeSkill", secondary: "someSkill" }
->{ primary: "SomeSkill", secondary: "someSkill" } : { primary?: string; secondary?: string; }
+>{ primary: "SomeSkill", secondary: "someSkill" } : { primary: string; secondary: string; }
 >primary : string
 >"SomeSkill" : "SomeSkill"
 >secondary : string

--- a/tests/cases/compiler/destructuringInitializerContextualTypeFromContext.ts
+++ b/tests/cases/compiler/destructuringInitializerContextualTypeFromContext.ts
@@ -1,0 +1,19 @@
+interface SFC<P = {}> {
+    (props: P & { children?: any }): any | null;
+}
+
+interface Props {
+    name: "Apollo" | "Artemis" | "Dionysus" | "Persephone";
+}
+
+const Parent: SFC<Props> = ({
+    children,
+    name = "Artemis",
+    ...props
+}) => Child({name, ...props});
+
+const Child: SFC<Props> = ({
+    children,
+    name = "Artemis",
+    ...props
+}) => `name: ${name} props: ${JSON.stringify(props)}`;

